src/Config/ConfigFactory.php

```diff
@@ -0,0 +1,130 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Psh\Config;
+
+use function array_merge;
+use function array_pop;
+use function array_shift;
+use function count;
+use function glob;
+use function is_dir;
+
+class ConfigFactory
+{
+    /**
+     * @var ConfigMerger
+     */
+    private $configMerger;
+
+    /**
+     * @var ConfigFileFinder
+     */
+    private $configFileFinder;
+
+    /**
+     * @var ConfigLoader[]
+     */
+    private $configLoaders;
+
+    /**
+     * @var ConfigLogger
+     */
+    private $configLogger;
+
+    public function __construct(
+        ConfigMerger $configMerger,
+        ConfigFileFinder $configFileFinder,
+        ConfigLogger $configLogger,
+        array $configLoaders
+    ) {
+        $this->configMerger = $configMerger;
+        $this->configFileFinder = $configFileFinder;
+        $this->configLoaders = $configLoaders;
+        $this->configLogger = $configLogger;
+    }
+
+    /**
+     * @return Config[]
+     */
+    public function gatherConfigs(
+        string $fromPath,
+        array $configFiles,
+        array $overwrittenConsts
+    ): array {
+        $configs = [];
+        $additionalConfigs = [[]];
+
+        foreach ($configFiles as $configFile) {
+            foreach ($this->configLoaders as $configLoader) {
+                if (!$configLoader->isSupported($configFile)) {
+                    continue;
+                }
+
+                $config = $configLoader
+                    ->load($configFile, $overwrittenConsts);
+
+                $additionalConfigs[] = $this
+                    ->loadImports($config, $fromPath, $overwrittenConsts);
+
+                $configs[] = $config;
+            }
+        }
+
+        if (!$configs) {
+            return [];
+        }
+
+        $config = $this->configMerger->mergeOverride(...$configs);
+
+        return array_merge([$config], ...array_merge(...$additionalConfigs));
+    }
+
+    private function loadImports(
```

Comment:
Shoudn't the ConfigLoader do this?

Reply:
The ConfigLoaders are only responsible for Loading a single file, maybe I can get the naming clearer?

I renamed the interface `ConfigFileLoader` and made the abstract a trait just for file handling. I think this reflects the usage better, doesn't it?